source/_posts/2015-03-26-creating-a-dynamic-single-page-app-for-our-genome-team-using-react.markdown

```diff
@@ -0,0 +1,297 @@
+```javascript
+updateArtwork: function(artwork_id, cb) {
+  // finds an artwork, passes it to callback (cb) to be mutated,
+  // sets the mutated artwork on state from the return value of
+  // the callback
+  var new_artwork = cb(this.state.artworks[artwork_id]);
+  var state_copy = this.state;
+  state_copy.artworks[artwork_id] = new_artwork;
+  this.setState(state_copy);
+},
+```
+
+That method is passed to child components, and when there is an update to an `artwork`, such as when it becomes selected, we invoke it to update all affected components:
+```javascript
+changeIsSelected: function (e) {
+  e.preventDefault();
+  var newSelectedState = !this.props.artwork.isSelected;
+  this.props.updateArtwork(this.props.artwork._id, function(artwork) {
+    artwork.isSelected = newSelectedState;
+    return artwork;
+  })
+},
+```
+
+React lets us define our components and interactions in a declarative style instead of stringing together possible transitions triggered by events. Before converting this app to React, we had many bugs around form submission and saving genome progress. However, by modeling state instead of UI transitions, we could easily track changes and save progress incrementally in the background without requiring a page refresh from the user.
+
+#### From CoffeeScript to React: Selecting Artworks
+
+In this app, genomers are able to 'select' artworks on the panel for the purposes of saving and conducting batch actions. In our initial implementation, clicking the 'select all' button would individually click each artwork and used global event listeners to change UI state:
+
+```coffee
+($ 'body').on 'click', '.artwork-selector', (evt) ->
+  container = $(this)
+  currentSlug = container.data('id')
+  artworkIdsElement = $('#selected_artwork_ids')
+  selectedArtworkIds = _.compact(_.uniq(artworkIdsElement.val().split(',')))
+  indexOfCurrentSlug = selectedArtworkIds.indexOf(currentSlug)
+  if selectedArtworkIds.indexOf(currentSlug) != -1
+    selectedArtworkIds.splice(indexOfCurrentSlug, 1)
+    container.text('Not selected')
+    container.removeClass('btn-purple')
+  else
+    selectedArtworkIds.push currentSlug
+    container.text('Selected')
+    container.addClass('btn-purple')
+  $('#selected_artwork_ids').val(selectedArtworkIds.join(','))
+  return false
+```
+
+With React, we store whether or not an artwork is selected as part of our state, and the appearance of elements results from this variable. We use [class sets](https://facebook.github.io/react/docs/class-name-manipulation.html) to dynamically alter styles such as button color. When the `selected` state changes, React re-renders all components that depend on that variable.
+
+```javascript
+var SelectedButton = React.createClass({
+  changeIsSelected: function (e) {
+    e.preventDefault();
+    var newSelectedState = !this.props.artwork.isSelected;
+    this.props.updateArtwork(this.props.artwork._id, function(artwork) {
```

Comment:
#minor and up to you, but it might be worth replacing this logic with a summary comment, or you could explain it (since it deals with a React paradigm, just not the one you're currently discussing)